system/mp/mpbase_mpi.cpp

```diff
@@ -0,0 +1,266 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#define mp_decl DECL_EXPORT
+#include "platform.h"
+#include "jlib.hpp"
+#include "jlog.hpp"
+
+#include "mpbase.hpp"
+#include "mpi_wrapper.hpp"
+
+static INode *MyNode=NULL;
+static INode *NullNode=NULL;
+
+
+IGroup *createIGroup(rank_t num,INode **nodes){
+    return new NodeGroup();
```

Comment:
I agree on on this as well that we should have a single implementation for IGroup. Or atleast a "template" version of it. 

It didn't feel right to have my own implementation of mpbase.cpp but since it is unavoidable, I duplicated the existing one with all the code that was common for MP/MPI and moved other implementation specific stuff to different source files (i.e. mpi_wrapper.cpp). Unfortunately this meant IGroup implementation and the createIGroup(...) function as well. (IMO ideally they should have being declared in mpcomm.hpp).

Reply:
Well we said better to copy and make progress, however, if it makes it less clear and harder to abstract the difference then it becomes harder to review and then make changes without the code diverging from the original.

Yeah, I'm not sure really need you're own IGroup implementation, but if you do and given the above, I think it would be better to rename the existing CGroup to CGroupBase defined in hpp, move the MP specific part into CGroup derived from CGroupBase and derive your own CGroupMPI.
Assuming that is relatively easy to achieve and I think it will clear some things up.

As I say should avoid it if can, but splitting like this will at least make it more obvious what's really required or not.

Reply:
That split was temporary. Just so we can get headway in implementation. But I think before going ahead with a base class for IGroup we should talk about the code that would be common for both Jsocket and MPI implementation that I had to duplicate because they are residing along the source of implementation specific code. Or else we might end up having this discussion again. wdyt?

Reply:
ok, sounds sensible, can discuss on slack.